devicetree: tree-wide: add nexus map properties for arduino headers

pabigot opened the pull request:
We need to be able to specify GPIO flags in devicetree without that preventing translation from the Arduino specifier to the host GPIO specifier.  Set up to ignore the low 6 bits of the flags field when matching the child specifier, and to copy those bits to the parent specifier.

This is a blocking capability for topic-gpio where we need to be able to specify the active level of signals from sensors connected through Arduino shields.

To preemptively answer a question: the low 6 bits are defined for devicetree in Linux although Zephyr currently does not use bit 3.  Bit 3 is used to signal suspend/resume/reset persistance, so may be needed in the future.

pabigot committed:
devicetree: tree-wide: add nexus map properties for arduino headers

We need to be able to specify GPIO flags in devicetree without that
preventing translation from the Arduino specifier to the host GPIO
specifier.  Set up to ignore the low 6 bits of the flags field when
matching the child specifier, and to copy those bits to the parent
specifier.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

erwango reviewed (commented):
Should we add some test in order this is not omitted when new connector type is added?
(https://github.com/zephyrproject-rtos/zephyr/pull/19751 for instance)

pabigot commented:
> Should we add some test in order this is not omitted when new connector type is added?
> (#19751 for instance)

That could perhaps be done in the devicetree scripting if it were possible to determine that the nexus map is associated with GPIOs.  IMO its lack shouldn't slow incorporation of this change.

ulfalizer commented:
> That could perhaps be done in the devicetree scripting if it were possible to determine that the nexus map is associated with GPIOs. IMO its lack shouldn't slow incorporation of this change.

`_check_dt()` in `edtlib.py` is meant for random devicetree checks like that btw, so could add something there if you want.

If it doesn't need to look at low-level devicetree stuff, it could be added to `gen_defines.py` instead.

erwango reviewed (changes requested):
I'm fine with the change, but it is a tricky one, so I'd like to see some documentation explaining the "trick" with some details.
Motivations:
- Doc is always welcome
- This is only described in V0.3 device tree spec, and it hasn't been released yet
- Even though v0.3 would be the current release, device tree spec explanations are abstruse

pabigot commented:
> I'm fine with the change, but it is a tricky one, so I'd like to see some documentation explaining the "trick" with some details.

Wonderful.

@ulfalizer since you maintain the devicetree documentation could you provide a patch with this that I can add to this PR?  If you can point me to it in a branch I can cherry-pick it into this.

pabigot committed:
doc: dts: describe nexus node magic required for shield gpio translation

Document why and how we use the devicetree nexus map properties to
preserve devicetree flag specifications.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
@erwango @ulfalizer @dbkinder please review the newly added documentation on this feature.

erwango commented on doc/guides/dts/gpio_nexus.inc:
```diff
@@ -0,0 +1,74 @@
+In Zephyr GPIO specifiers generally have two parameters (indicated by
+``#gpio-cells = <2>``): the pin number and a set of flags.  The low 6
+bits of the flags correspond to features that can be configured in
+devicetree.  In some cases it's necessary to use a non-zero flag value
```
I don't know if there is a kind of bitmap drawing available we can point to but it would be nice.
(To do later since it is a different branch)

erwango reviewed (approved):
Much clearer than dts specification!

ulfalizer reviewed (approved):
Sorry, didn't notice the update until now.

Looks good to me.

pabigot commented:
@dbkinder Could you please check the documentation addition to this for acceptability?  This PR blocks work on driver conversion to the new gpio API.  Thanks.

dbkinder reviewed (approved):
While I'd prefer the gpio_nexus.inc content to be put in the rst file directly, it's OK as is.

pabigot commented on doc/guides/dts/index.rst:
```diff
@@ -712,4 +712,5 @@ generated in the same directory concatenating all board-related fixup files.
 The include file ``include/generated_dts_board.h`` includes both these generated
 files, giving Zephyr C source files access to the board's devicetree information.
 
+.. include:: gpio_nexus.inc
```
OK, I'll integrate it into the file.  I need to rebase and add a shippable override to get CI to pass anyway.

pabigot commented:
@galak all tests pass via https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/54062/summary/console

I've now removed the shippable override.  This should be good for merge.
